### PR TITLE
Compositions of functions

### PR DESCRIPTION
**Pull request by avekens**
Recently, I had a discussion with Gérard Lang about compositions of functions. The results, mainly suggested by Gérard, are contained in this PR. The main results are:

* ~fcores: Every composite function ` ( G o. F ) ` can be written as composition of restrictions of the composed functions (to their minimum domains)
* ~fcoresf1,  ~fcoresfo:  If a composition is injective/surjective, then the minimum restrictions of its components are injective/surjective
* ~f1cof1b, ~focofob, ~f1ocof1ob: The composition ` ( G o. F ) ` is surjective/injective/bijective iff ` F ` and ` G ` are both surjective/injective/bijective (if the range of ` F ` equals the domain of ` G ` )

which I put into my mathbox (I do not mind if they were moved to main).

To prove these, I needed some auxiliary theorems, which are mostly generalizations of already existing theorems. Therefore, I put them directly into main, because I could shorten the proofs of the former theorems then.

Details:
* ~sylibda, ~ineqcom, ~ineqcomi, ~freld, ~fco3 (renamed) moved from mathboxes to main; proofs for ~funcofd (formerly ~fco3) and ~smfco shortened
* new theorem ~cnvimassrndm (generalization of ~cnvimarndm), proofs for ~cnvimarndm, ~fimacnv shortened
* new theorem ~ffrnb (generalization of ~ffrn)
* new theorems ~fncofn, ~fcof, ~f1cof1 (generalizations of ~fnco, ~fco, f1co), proofs shortened
* new theorems ~cnvimainrn , ~rescnvimafod added

AV's mathbox:
* comment for ~fsetabsnop  revised (as proposed by GL)
* new theorems ~fcores, ~fcoresf1,  ~fcoresfo and lemmas added
* new theorems ~f1cof1b, ~focofob, ~f1ocof1ob added

### CODE REVIEW

File: set.mm
Comment (icecream17) [inline set.mm:5478]: just as in #4220, sylibda should've been named sylbida...

File: set.mm
Comment (icecream17) [inline set.mm:38557]:
```suggestion
      ( cin incom eqtri ) BAEABECBAFDG $.
```
TODO-SN: nonblocking

**Review by icecream17 — APPROVED** [review]
nice results! (f1cof1b, focofob, and f1ocof1ob are generalizations of f1co, foco, and f1oco respectively)

File: set.mm
Comment (benjub) [inline set.mm:61420]: The biconditional
```
$( Characterization of a function with domain and codomain (essentially using that the
   range is always included in the codomain).
$p |- ( F : A --> B <-> ( F : A --> ran F /\ ran F C_ B ) ) $=
```
seems to offer a more precise formulation.

File: set.mm
Comment (benjub) [inline set.mm:61347]:
```suggestion
    $( A function with domain and codomain is a relation.  (Contributed by Glauco Siliprandi,
```
That's our (somewhat arbitrary) convention to distinguish our various definitions:
- Fun F : a function
- F Fn A : a function with domain
- F : A --> B : a function with domain and codomain

File: set.mm
Comment (benjub) [inline set.mm:60882]:
```suggestion
  $( Composition of a function with domain and a function as a function with domain.
```

File: set.mm
Comment (benjub) [inline set.mm:60889]:
```suggestion
  $( Composition of two functions with domains as a function with domain.  (Contributed by NM, 22-May-2006.)  (Proof
```

File: set.mm
Comment (benjub) [inline set.mm:38547]:
```suggestion
  $( Two ways of expressing that two classes have a given intersection.  This is
     often used when that given intersection is the empty set, in which case the
     statement displays two ways of expressing that two classes are disjoint.
     (Contributed by Peter
```

File: set.mm
Comment (benjub) [inline set.mm:38554]:
```suggestion
    $( If the intersection of two classes is a given class, then their intersection
       when swapped is that same class.  Inference form of
```

File: set.mm
Comment (benjub) [inline set.mm:61443]:
```suggestion
  $( A direct image of a function with domain and codomain is a subset of its codomain.  (Contributed by Glauco
```

File: set.mm
Comment (benjub) [inline set.mm:61450]:
```suggestion
  $( The preimage of the codomain of a function with domain and codomain is the function's domain.
     (Contributed by FL, 25-Jan-2007.)  (Proof shortened by AV,
     20-Sep-2024.) $)
```

File: set.mm
Comment (benjub) [inline set.mm:61463]:
```suggestion
  $( Composition of two functions with domain and codomain as a function with domain and codomain.  (Contributed by NM, 29-Aug-1999.)  (Proof
```

File: set.mm
Comment (benjub) [inline set.mm:61456]:
```suggestion
  $( Composition of a function with domain and codomain and a
     function as a function with domain and codomain.  Generalization of ~ fco .  (Contributed by AV, 18-Sep-2024.) $)
```

File: set.mm
Comment (benjub) [inline set.mm:61535]:
```suggestion
    $( Composition of two functions as a function with domain and codomain.  (Contributed by
```

File: set.mm
Comment (benjub) [inline set.mm:61937]:
```suggestion
  $( Composition of two one-to-one functions.  Generalization of ~ f1co .
```

File: set.mm
Comment (benjub) [inline set.mm:61946]:
```suggestion
  $( Composition of two one-to-one functions when the codomain of the first matches the domain of the second.  Exercise 30 of [TakeutiZaring]
```

File: set.mm
Comment (benjub) [inline set.mm:748805]: It would be clearer if you rename X to Y and Y to X everywhere.

File: set.mm
Comment (benjub) [inline set.mm:748819]: From that point on, you have hypotheses with `C` and hypotheses with `dom G` (in the definition of `E`), which yields a mix which makes statements look less elegant.

File: set.mm
Comment (benjub) [inline set.mm:748861]: "minimum restriction" --> "restriction to the smallest..."

(Because to "minimally restrict" is to not restrict at all.)

File: set.mm
Comment (benjub) [inline set.mm:748923]: Here and below, the condition in the parentheses is essential, so I would remove the parentheses and put it first:
> If the range of ` F ` equals the domain of ` G ` , then the composition...

**Review by benjub — COMMENTED** [review]
Since there are many similar statements in these sections around the definition**s** of functions, restrictions, etc., I think it is important to be more uniform and systematic in the comments, hence my suggestions. Ideally, we would also review existing comments and copy them to iset.mm too.

File: set.mm
Comment (avekens) [reply set.mm:5478]: I will fix that, bit it seems to result in a merge conflict anyway...

File: set.mm
Comment (avekens) [reply set.mm:38557]: Done.

File: set.mm
Comment (avekens) [reply set.mm]: Done

**Comment by avekens** [comment]
> Since there are many similar statements in these sections around the definition**s** of functions, restrictions, etc., I think it is important to be more uniform and systematic in the comments, hence my suggestions. Ideally, we would also review existing comments and copy them to iset.mm too.

I am not happy with the structure of the whole subsection  2.3.15  "Functions": it's huge (about 800 theorems) and it is always difficult to find theorems or to place new theorems. IMHO, it should be subdivided into subsubsections. However, there could be two different approaches: Either to create a subsubsection for each (of the 7) definitions, or to  create a subsubsection for " basic theorems and then a subsubsection for each (of the 8) definitions of subsection  2.3.10  "Relations" (relation, domain, range, etc.), relating these former concepts to the (new) definitions of functions. Of course both approaches could be combined, but this would result in 8x8=64 subsubsections, which is definitely to much. 7-10 subsubsections (containing about 100 theorems each) seem to be adequate.

Maybe a first step could be to split 2.3.15  "Functions" into two subsubsections 2.3.15.1 "Plain functions" (with definitions df-fun, df-fn, df-f and related theorems) and 2.3.15.2 "Special functions" (with definitions df-f1, df-fo, df-f1o,  df-fv, df-isom) and related theorems).

File: set.mm
Comment (avekens) [reply set.mm:61443]: I will change the comment to "The image of a class under a function with domain and codomain is a subset of its codomain." I do not now if the term "direct image" is adequate here, at least it is not defined at this place.

File: set.mm
Comment (avekens) [reply set.mm:748819]: That's true, I observed this, too. I will add a TODO here and will fix this in a subsequent PR.

File: set.mm
Comment (avekens) [reply set.mm]: I will add this theorem as ~ffrnb, and I will revise the current ~ffrnb to become ~ffrnbd in deduction form.

File: set.mm
Comment (avekens) [reply set.mm]: I tried to use these specific wordings as much as possible/meaningful in my new theorems. However,  I think it is not always necessary to use the full name, e.g., if the word "codomain" appears already within the comment: "The codomain of a function is .." should be sufficient compared with "The codomain of a function with domain and codomain is .."

File: set.mm
Comment (avekens) [reply set.mm]: You are right: th current Y corresponds to F, and the current X corresponds to G.  I will add a TODO here and will fix this in a subsequent PR.

File: set.mm
Comment (avekens) [reply set.mm]: I will replace "mapping" by "function" here, but this is a case in which "with domain and codomain" is not necessary, see my comment above.

File: set.mm
Comment (avekens) [reply set.mm]: I think the "iff" statement is the most important part, and usually the range of ` F ` equals the domain of ` G `.

File: set.mm
Comment (zwang123) [reply set.mm]: Oops I fixed that too. Don't know if there will be a conflict...

Comment (avekens) [reply set.mm]: Don't worry, if there are merge conflicts, they can be solved easily (in this case).

File: set.mm
Comment (benjub) [inline set.mm:61429]:
```suggestion
     ~ ffrn .  (Contributed by BJ and AV, 21-Sep-2024.) $)
```
Or only BJ, or only AV, as you prefer: you did the work and I only reformulated.

File: set.mm
Comment (benjub) [reply set.mm:61443]: As you prefer.

File: set.mm
Comment (benjub) [reply set.mm]: No, that condition is essential and should be placed first, and not in parentheses. When I did read your sentence the first time, I really startled before I could even read the parenthesis.

File: set.mm
Comment (benjub) [reply set.mm]: You're right, I think you described a good balance.

**Comment by benjub** [comment]
> > Since there are many similar statements in these sections around the definition**s** of functions, restrictions, etc., I think it is important to be more uniform and systematic in the comments, hence my suggestions. Ideally, we would also review existing comments and copy them to iset.mm too.
> 
> I am not happy with the structure of the whole subsection 2.3.15 "Functions": it's huge (about 800 theorems) and it is always difficult to find theorems or to place new theorems. IMHO, it should be subdivided into subsubsections. However, there could be two different approaches: Either to create a subsubsection for each (of the 7) definitions, or to create a subsubsection for " basic theorems and then a subsubsection for each (of the 8) definitions of subsection 2.3.10 "Relations" (relation, domain, range, etc.), relating these former concepts to the (new) definitions of functions. Of course both approaches could be combined, but this would result in 8x8=64 subsubsections, which is definitely to much. 7-10 subsubsections (containing about 100 theorems each) seem to be adequate.
> 
> Maybe a first step could be to split 2.3.15 "Functions" into two subsubsections 2.3.15.1 "Plain functions" (with definitions df-fun, df-fn, df-f and related theorems) and 2.3.15.2 "Special functions" (with definitions df-f1, df-fo, df-f1o, df-fv, df-isom) and related theorems).

I couldn't agree more. It is actually that behemoth section that made me open the issue #2861. I still think sectioning by definition will provide a better structuring, easier to navigate. (And "basic theorems"/"less basic" is too subjective.) Let's move the discussion there for the action plan.

File: set.mm
Comment (avekens) [reply set.mm:61429]: The proof wasn't much work - therefore no change of contributor() necessary.

File: set.mm
Comment (avekens) [reply set.mm]: OK, I revised the comments accordingly.

File: set.mm
Comment (avekens) [reply set.mm]: Done already in this PR.

File: set.mm
Comment (avekens) [reply set.mm]: Done already in this PR.

**Review by benjub — APPROVED** [review]

**Comment by benjub** [comment]
@avekens Do you see how to fix the pipeline fail ? (Beware: I merged `develop`, so you'll have to `git pull` this branch first, or else you ignore this and work locally without pulling this one, and then you force-push (so you override my latest commit). I can also do it if you prefer: just ping me here.

**Comment by avekens** [comment]
The problem is that the theorem sylbida (formerly sylibda) was revised/moved in two PR in parallel.
I'll try to fix this within the next two days.

**Comment by benjub** [comment]
I fixed the problem by making one additional commit and making MR #4238. I need one approval there (after checking that the list of commits is identical except that there is one additional one) to merge it and close this one.

**Comment by avekens** [comment]
This PR can be closed after #4238 is merged.

**Comment by benjub** [comment]
Closing as supserseded by #4238.

**Pull request closed without merging**
